tests/Unit/NumericalAlgorithms/DiscontinuousGalerkin/Test_Mass.cpp

```diff
@@ -0,0 +1,108 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/Tensor/EagerMath/Determinant.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/CoordinateMaps/Affine.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.tpp"
+#include "Domain/CoordinateMaps/ProductMaps.hpp"
+#include "Domain/CoordinateMaps/ProductMaps.tpp"
+#include "Domain/LogicalCoordinates.hpp"
+#include "NumericalAlgorithms/DiscontinuousGalerkin/Mass.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace dg {
+namespace {
+
+template <size_t Dim, typename... CoordMaps>
+void test_mass(const Mesh<Dim>& mesh,
+               const domain::CoordinateMap<Frame::Logical, Frame::Inertial,
+                                           CoordMaps...>& coordinate_map,
+               const DataVector& scalar_field,
+               const DataVector& expected_massive_scalar_field) noexcept {
+  const size_t num_grid_points = mesh.number_of_grid_points();
+  const auto logical_coords = logical_coordinates(mesh);
+  const auto jacobian = coordinate_map.jacobian(logical_coords);
+  const auto det_jacobian = determinant(jacobian);
+  {
+    INFO("Test with DataVector");
+    auto result = scalar_field;
+    result *= get(det_jacobian);
+    apply_mass(make_not_null(&result), mesh);
+    CHECK_ITERABLE_APPROX(result, expected_massive_scalar_field);
+  }
+  {
+    INFO("Test with Variables");
+    using tag1 = ::Tags::TempScalar<0>;
+    using tag2 = ::Tags::TempScalar<1>;
+    Variables<tmpl::list<tag1, tag2>> vars{num_grid_points};
+    get<tag1>(vars) = Scalar<DataVector>(scalar_field);
+    get<tag2>(vars) = Scalar<DataVector>(scalar_field);
+    vars *= get(det_jacobian);
+    apply_mass(make_not_null(&vars), mesh);
+    CHECK_ITERABLE_APPROX(get(get<tag1>(vars)), expected_massive_scalar_field);
+    CHECK_ITERABLE_APPROX(get(get<tag2>(vars)), expected_massive_scalar_field);
+  }
+}
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Numerical.DiscontinuousGalerkin.Mass",
+                  "[NumericalAlgorithms][Unit]") {
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine2D = domain::CoordinateMaps::ProductOf2Maps<Affine, Affine>;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+
+  {
+    INFO("1D");
+    const Mesh<1> mesh{
+        {{4}}, Spectral::Basis::Legendre, Spectral::Quadrature::GaussLobatto};
+    const auto coord_map =
+        domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(
+            Affine{-1.0, 1.0, -0.3, 0.7});
+    test_mass(mesh, coord_map, {1., 2., 3., 4.},
```

Comment:
is there a way to compute the mass matrix from the nodal to modal matrices or something? having a bunch of magic numbers in a test file isn't fabulous :shrug:

Reply:
Added a test with the mass matrix computed from the Vandermonde matrix. Note that it's only equivalent for Gauss quadrature, because of the approximation made for Gauss-Lobatto quadrature. Therefore I kept the hard-coded numbers around for Gauss-Lobatto quadrature. They are just the data divided by the quadrature weights, so I added a comment.